gematria/datasets/find_accessed_addrs_exegesis.cc

```diff
@@ -0,0 +1,152 @@
+
+  if (!RunnerOrErr) return RunnerOrErr.takeError();
+
+  std::unique_ptr<const SnippetRepetitor> Repetitor_ =
+      SnippetRepetitor::Create(Benchmark::RepetitionModeE::Duplicate, State_);
+
+  return std::unique_ptr<ExegesisAnnotator>(new ExegesisAnnotator(
+      ArchSupport_, State_, std::move(*RunnerOrErr), std::move(Repetitor_)));
+}
+
+Expected<AccessedAddrs> ExegesisAnnotator::FindAccessedAddrs(
+    ArrayRef<uint8_t> BasicBlock) {
+  Expected<std::vector<DisassembledInstruction>> DisInstructions =
+      DisassembleAllInstructions(
+          ArchSupport.mc_disassembler(), ArchSupport.mc_instr_info(),
+          ArchSupport.mc_register_info(), ArchSupport.mc_subtarget_info(),
+          *MCPrinter, 0, BasicBlock);
+
+  if (!DisInstructions) return DisInstructions.takeError();
+
+  std::vector<MCInst> Instructions;
+  Instructions.reserve(DisInstructions->size());
+
+  for (const auto &DisInstruction : *DisInstructions)
+    Instructions.push_back(DisInstruction.mc_inst);
+
+  AccessedAddrs MemAnnotations;
+  MemAnnotations.code_location = 0;
+  MemAnnotations.block_size = 4096;
+
+  BenchmarkCode BenchCode;
+  BenchCode.Key.Instructions = Instructions;
+
+  MemoryValue MemVal;
+  MemVal.Value = APInt(64, 0x12345600);
+  MemVal.Index = 0;
+  MemVal.SizeBytes = 4096;
+
+  BenchCode.Key.MemoryValues["memdef1"] = MemVal;
+
+  const llvm::MCRegisterInfo &MRI = ArchSupport.mc_register_info();
+
+  for (unsigned i = 0;
+       i < MRI.getRegClass(X86::GR64_NOREX2RegClassID).getNumRegs(); ++i) {
+    RegisterValue RegVal;
+    RegVal.Register =
+        MRI.getRegClass(X86::GR64_NOREX2RegClassID).getRegister(i);
+    RegVal.Value = APInt(64, 0x12345600);
+    BenchCode.Key.RegisterInitialValues.push_back(RegVal);
+  }
+
+  for (unsigned i = 0; i < MRI.getRegClass(X86::VR128RegClassID).getNumRegs();
+       ++i) {
+    RegisterValue RegVal;
+    RegVal.Register = MRI.getRegClass(X86::VR128RegClassID).getRegister(i);
+    RegVal.Value = APInt(128, 0x12345600);
+    BenchCode.Key.RegisterInitialValues.push_back(RegVal);
+  }
+
+  while (true) {
+    std::unique_ptr<const SnippetRepetitor> SR =
+        SnippetRepetitor::Create(Benchmark::RepetitionModeE::Duplicate, State);
+    Expected<BenchmarkRunner::RunnableConfiguration> RCOrErr =
+        Runner->getRunnableConfiguration(BenchCode, 10000, 0, *SR);
+
+    if (!RCOrErr) return RCOrErr.takeError();
+
+    BenchmarkRunner::RunnableConfiguration &RC = *RCOrErr;
+
+    auto BenchmarkResults = Runner->runConfiguration(std::move(RC), {}, true);
+
+    // If we don't have any errors executing the snippet, we executed the
+    // snippet successfully and thus have all the needed memory annotations.
+    if (BenchmarkResults) break;
+
+    auto ResultError = BenchmarkResults.takeError();
+
+    if (!ResultError.isA<SnippetCrash>()) return std::move(ResultError);
+
+    handleAllErrors(std::move(ResultError), [&](SnippetCrash &CrashInfo) {
+      if (CrashInfo.GetCrashAddress() == 0) return;
```

Comment:
This probably needs to turn into an actual crash. We can map pages under the 64kb limit in some contexts, but it's generally not an encouraged practice. Will have to see the practical implications though.

This case currently happens transiently due to the restartable sequence not getting disabled in glibc due to some weird preprocessor stuff happening in Bazel. Will investigate it and I think it will probably end up needing a patch on the LLVM side. Hacky fix that I've been experimenting with is to adjust some of the preprocessor logic in the head of `llvm/tools/llvm-exegesis/lib/BenchmarkRunner.cpp`.

Reply:
Ah. It's because `HAVE_BUILTIN_THREAD_POINTER` is necessary for the preprocessor check to pass and is defined by CMake, but isn't defined by the Bazel build. Will post a patch to get that fixed.

Reply:
Fixed in https://github.com/llvm/llvm-project/pull/74574.